operator/pkg/controllers/etcd/etcd.go

```diff
@@ -0,0 +1,346 @@
+			Labels:    etcdLabelFor(controlPlane.ClusterName()),
+			OwnerReferences: []metav1.OwnerReference{{
+				APIVersion: controlPlane.APIVersion,
+				Name:       controlPlane.Name,
+				Kind:       controlPlane.Kind,
+				UID:        controlPlane.UID,
+			}},
+		},
+		Spec: v1.ServiceSpec{
+			Ports: []v1.ServicePort{{
+				Port: 2380,
+				Name: etcdServerPortNameFor(controlPlane.ClusterName()),
+			}, {
+				Port: 2379,
+				Name: etcdClientPortNameFor(controlPlane.ClusterName()),
+			}},
+			ClusterIP: "None",
+			Selector:  etcdLabelFor(controlPlane.ClusterName()),
+		},
+	}
+	if err := e.kubeClient.Create(ctx, svc); err != nil {
+		return fmt.Errorf("creating service %s/%s, %w", svc.Namespace, svc.Name, err)
+	}
+	zap.S().Infof("[%s] created service %s", controlPlane.ClusterName(),
+		etcdServiceNameFor(controlPlane.ClusterName()))
+	return nil
+}
+
+func (e *Provider) getService(ctx context.Context, name types.NamespacedName) (*v1.Service, error) {
+	result := &v1.Service{}
+	if err := e.kubeClient.Get(ctx, name, result); err != nil {
+		return nil, fmt.Errorf("getting service %v, %w", name, err)
+	}
+	return result, nil
+}
+
+func (e *Provider) createETCDCerts(ctx context.Context, controlPlane *v1alpha1.ControlPlane) error {
+	// get CA key and cert
+	caCert, caKey, err := e.createCertAndKeyIfNotFound(ctx, controlPlane, etcdRootCACertConfig(controlPlane))
+	if err != nil {
+		return fmt.Errorf("creating root CA for cluster %v, %w", controlPlane.ClusterName(), err)
+	}
+	// generate etcd server, peer, healthcheck and etcdAPIClient certs
+	for _, certConfig := range certListFor(controlPlane, caCert, caKey) {
+		if _, _, err := e.createCertAndKeyIfNotFound(ctx, controlPlane, certConfig); err != nil {
+			return fmt.Errorf("creating certs and key name %s, %w,", certConfig.name, err)
+		}
+	}
+	return nil
+}
+
+// createCertAndKeyIfNotFound checks with management server for a root CA secret, if it exists return the cert and key, else
+// it will generate a new root ca and store as a secret in management server
+func (e *Provider) createCertAndKeyIfNotFound(ctx context.Context, controlPlane *v1alpha1.ControlPlane, certConfig *CertConfig) (*x509.Certificate, crypto.Signer, error) {
+	var cert *x509.Certificate
+	var key crypto.Signer
+	secret, err := e.getSecret(ctx, controlPlane.NamespaceName(), certConfig.name)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			cert, key, err = e.createCertAndKeySecret(ctx, controlPlane, certConfig)
+			if err != nil {
+				return nil, nil, err
+			}
+			return cert, key, err
+		}
+		return nil, nil, err
+	}
+	certs, err := certutil.ParseCertsPEM(secret.Data["tls.crt"])
+	if err != nil {
+		return nil, nil, err
+	}
+	parsedKey, err := keyutil.ParsePrivateKeyPEM(secret.Data["tls.key"])
+	if err != nil {
+		return nil, nil, err
+	}
+	return certs[0], parsedKey.(crypto.Signer), err
+}
+
+func (e *Provider) getSecret(ctx context.Context, namespace, objName string) (*v1.Secret, error) {
```

Comment:
I'd avoid helpers like this, since they don't differentiate over the implementation.